Switch windows workflow to run on Azure hosted ARC runner (#18859)

Eliasj42 opened the pull request:
Progress on https://github.com/iree-org/iree/issues/18813

workflow running: https://github.com/iree-org/iree/actions/runs/11448943622/job/31853515374

---------

Eliasj42 committed:
Switch windows workflow to run on Azure hosted ARC runner (#18859)

Progress on https://github.com/iree-org/iree/issues/18813

---------

Signed-off-by: Elias Joseph <eljoseph@amd.com>
Co-authored-by: Elias Joseph <eljoseph@amd.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

marbre committed:
[docs] Update iree-compiler install instructions (#18861)

Instructions to install iree-runtime are provided separately.

Signed-off-by: Elias Joseph <eljoseph@amd.com>

Max191 committed:
Integrate llvm-project @864902e9b4d8bc6d3f0852d5c475e3dc97dd8335 (#18843)

bump llvm-project to
https://github.com/llvm/llvm-project/commit/864902e9b4d8bc6d3f0852d5c475e3dc97dd8335

Main changes in IREE (causing test updates):
- New forall loop canonicalizations are added which fold induction
variables. This can canonicalilze away some index computation, and fold
some extract_slice ops, so a few pipeline tests are changed.
- A new option for `hoistRedundantVectorTransfers` (`verifyNonZeroTrip`)
is set, which prevents hoisting from possibly zero trip loops. This
option is flipped in this integrate because it fixes a bug exposed by
forall loop canonicalizations.

---------

Signed-off-by: Max Dawkins <max.dawkins@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

ScottTodd commented on .github/workflows/ci_windows_x64_msvc.yml:
```diff
@@ -22,19 +22,22 @@ concurrency:
+          Get-PSDrive | Format-Table Name, Used, Free
+          Test-Path C:\mnt\azure
+          Get-ChildItem -Path C:\mnt\azure
+          fsutil volume diskfree C:\mnt\azure
```
All generic (non-project-specific) setup should be moved to the Dockerfile. I'm less concerned about steps taking up to 1 minute, but the 15 minute MSVC install will need to be moved to ahead of time. Our target for presubmit should be 10-20 minutes per run, including the source build (warm cache). We can budget for 5 minutes or so of overhead (git checkout, installing deps, etc.).

ScottTodd commented on .github/workflows/ci_windows_x64_msvc.yml:
```diff
@@ -22,19 +22,22 @@ concurrency:
-    defaults:
-      run:
-        shell: bash
+    runs-on: arc-runner-set
```
How many cores is this configured with right now? https://github.com/iree-org/iree/actions/runs/11448943622/job/31853515374 is making slow progress.

@saienduri said that we might only have quota for small runners, but we need to make a case for larger runners based on utilization or something? If so, that's silly - we know that we need 32 (or even 64/96) core runners and we already have data to back that up. We _could_ switch the current nightly build to these new runners if it is necessary to make a case for quota increases, but we're currently paying $0 for free runners that take 4h30m. Paying for runners that are similarly slow doesn't make sense to me :P

cc @amd-chrissosa

saienduri replied:
Currently, this cluster is using instances with 8 cpu cores

ScottTodd commented on .github/workflows/ci_windows_x64_msvc.yml:
```diff
@@ -44,9 +47,37 @@ jobs:
+      - name: Add Bash to PATH
+        run: |
+          echo "Adding Bash to PATH"
+          echo "C:\Program Files\Git\bin" >> $Env:GITHUB_PATH
```
(All of these steps should move to the Dockerfile too. Some amount of workflow scripting is okay if it unblocks these new runners faster... _if the overhead is low enough_)

ScottTodd commented on .github/workflows/ci_windows_x64_msvc.yml:
```diff
@@ -44,9 +47,37 @@ jobs:
       - name: "Testing IREE"
-        run: ./build_tools/cmake/ctest_all.sh "${BUILD_DIR}"
+        run: bash ./build_tools/cmake/ctest_all.sh "/c/mnt/azure/build-windows"
+      - name: "Clean up build dir"
+        run: Remove-Item -Path "C:\mnt\azure\build-windows" -Recurse -Force
```
Is this necessary? The GitHub runner should do some automatic cleanup already. We can tolerate some scripting like this if it unblocks other work, but this does increase workflow complexity. This also fails to clean up the directory if the workflow fails or is cancelled (it would need a `if(always)`... but even that can fail to run if the runner goes offline before running that step).

Maybe keep the build directory in the github workspace? Why did you put things under `C:\mnt\azure\`?

saienduri replied:
For some reason, the default storage drive is hard set to less space than needed and flipping different switches/requesting more memory wasn't able to resolve it for the pod instances. So, we decided to use a storage mount that we have full control over

ScottTodd commented on .github/workflows/ci_windows_x64_msvc.yml:
```diff
@@ -44,9 +47,37 @@ jobs:
+        run: |
+          echo "Adding Bash to PATH"
+          echo "C:\Program Files\Git\bin" >> $Env:GITHUB_PATH
+      - name: "Building Iree"
+        run: bash ./build_tools/cmake/build_all.sh "/c/mnt/azure/build-windows"
```
Build error: https://github.com/iree-org/iree/actions/runs/11448943622/job/31853515374#step:13:8554

```
[7708/8119] Linking CXX shared library tools\IREECompiler.dll
FAILED: tools/IREECompiler.dll lib/IREECompiler.lib 
C:\Windows\system32\cmd.exe /C "cd . && "C:\Program Files\Microsoft Visual Studio\2022\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -E vs_link_dll --intdir=compiler\src\iree\compiler\API\CMakeFiles\iree_compiler_API_SharedImpl.dir --rc="C:\Program Files (x86)\Windows Kits\10\bin\10.0.22621.0\x64\rc.exe" --mt="C:\Program Files (x86)\Windows Kits\10\bin\10.0.22621.0\x64\mt.exe" --manifests  -- "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.41.34120\bin\Hostx64\x64\link.exe" /nologo @CMakeFiles\iree_compiler_API_SharedImpl.rsp  /out:tools\IREECompiler.dll /implib:lib\IREECompiler.lib /pdb:tools\IREECompiler.pdb /dll /version:0.0 /machine:x64 -fuse-ld=lld /debug /INCREMENTAL  -natvis:C:/home/runner/_work/iree/iree/runtime/iree.natvis && cd ."
LINK Pass 1: command "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.41.34120\bin\Hostx64\x64\link.exe /nologo @CMakeFiles\iree_compiler_API_SharedImpl.rsp /out:tools\IREECompiler.dll /implib:lib\IREECompiler.lib /pdb:tools\IREECompiler.pdb /dll /version:0.0 /machine:x64 -fuse-ld=lld /debug /INCREMENTAL -natvis:C:/home/runner/_work/iree/iree/runtime/iree.natvis /MANIFEST /MANIFESTFILE:compiler\src\iree\compiler\API\CMakeFiles\iree_compiler_API_SharedImpl.dir/intermediate.manifest compiler\src\iree\compiler\API\CMakeFiles\iree_compiler_API_SharedImpl.dir/manifest.res" failed (exit code 1140) with the following output:
LINK : warning LNK4044: unrecognized option '/fuse-ld=lld'; ignored

LINK : warning LNK4044: unrecognized option '/fuse-ld=lld'; ignored

   Creating library lib\IREECompiler.lib and object lib\IREECompiler.exp

LINK : fatal error LNK1318: Unexpected PDB error; LIMIT (12) 'C:\mnt\azure\build-windows\tools\IREECompiler.pdb'
```

Hmmm... could be multiple reasons for that. First thing I'd suspect is out of disk / RAM.

Eliasj42 replied:
Hi, sorry I've been neglecting this page while troubleshooting the above issue.  I fixed it, the only remaining issue is a certification error 

ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1007)

https://github.com/iree-org/iree/actions/runs/11469526800/job/31916920407

For some reason when I install msvc in the Docker container, the workflow can't find it.

ScottTodd reviewed (commented):
BTW this is great progress, thanks so much! Getting the runners set up with the general set of software installed is a huge milestone.

Commit:
removed msvc setup

Signed-off-by: Elias Joseph <eljoseph@amd.com>

Commit:
Revert "removed msvc setup"

This reverts commit 06aaf4a692eda59973b72bd40ffc4d11754e3574.

Signed-off-by: Elias Joseph <eljoseph@amd.com>

Commit:
removed msvc requirements

Signed-off-by: Elias Joseph <eljoseph@amd.com>

c-rhodes committed:
[CPU] Limit vectorization tile sizes for SVE (#18846)

This prevents large vector sizes. A regression test is included with a
linalg.pooling_nchw_max operation that currently fails to compile with

error: One or more operations with large vector sizes (8192 bytes) were
found:

when SVE is enabled, even though SVE isn't used.

---------

Signed-off-by: Cullen Rhodes <cullen.rhodes@arm.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

Groverkss committed:
[Attention] Only clamp attention for low precision types (#18848)

Post-softmax, the range of output is between 0, 1. For low-precision
types (like fp8), we scale the output range to be between 0, fpMax, so
we can use more of the fp range. For higher precision types like fp16,
we don't do this, so we don't need to clamp the range.

Signed-off-by: Elias Joseph <eljoseph@amd.com>

bjacob committed:
GPU data tiling: query the target's list of MMA intrinsics. Add FP8 test. (#18862)

The current code had its own list of MFMA intrinsics that we can use,
then checked that against the target. Flipping this around, we can
simply query the list from the target.

The only subtlety is that the target may support multiple intrinsics for
a given combination of element types, in which case we have to choose
one.

This PR also changes `std::optional<Attr>` to just `Attr` since a
default-constructed `Attr` is null-ish, there is no need for a second
null-value.

The heuristic added in this PR is designed to match the existing choices
so that the tests don't need to change; these existing choices are also
what maximizes some microbenchmark performance, but we have known that
they may be counterproductive in real scenarios where the bottleneck is
power.

The test gains a FP8 testcase, and some renaming to simplify function
names (which had become a lie in some testcases).

---------

Signed-off-by: Benoit Jacob <jacob.benoit.1@gmail.com>
Co-authored-by: Quinn Dawkins <quinn.dawkins@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

Groverkss committed:
[Codegen] Allow multiple reduction dimensions in VectorDistribution (#18868)

This PR adds support for multiple k dimensions in VectorDistribution
contract codegen.

Signed-off-by: Elias Joseph <eljoseph@amd.com>

vivekkhandelwal1 committed:
Bump torch-mlir to 140cad5 and update TorchOnnxToTorch conversion pipeline (#18867)

This commit bumps torch-mlir to
https://github.com/llvm/torch-mlir/commit/140cad5659bb779bb1f5de1888566db5b5d21236.

This commit also replaces the `TorchOnnxToTorchPass` used to convert
torch-onnx IR to torch IR with `TorchOnnxToTorchBackendPipeline` which
was introduced here:
https://github.com/llvm/torch-mlir/commit/fa4794dae2057876ec8ad2a6464e2668f6a2ea0c.
This pipeline consists of passes like `TorchOnnxToTorch`,
`ScalarizeShapes`, `DecomposeComplexOps`, `ShapeRefinementPipeline`,
etc., to convert the torch-onnx IR to torch IR consistent with the other
lowering paths for torch backend IR.

With the changes made in this PR, some of the tests which were failing
earlier during compilation now passes compilation, some of them even
passes inference, while there are some tests which were passing earlier
are now failing. The list is as follows:
- For
tests/external/iree-test-suites/onnx_ops/onnx_ops_cpu_llvm_sync.json:
  a.) Earlier failing compilation, now passing inference:
    - onnx/node/generated/test_convtranspose_kernel_shape
    - onnx/node/generated/test_einsum_sum

b.) Earlier failing compilation, now passing compilation but failing
inference
    - onnx/node/generated/test_convtranspose_output_shape

  c.) Earlier passing but now failing inference
    - onnx/node/generated/test_scan_sum
    - onnx/node/generated/test_scan9_sum

- For
tests/external/iree-test-suites/onnx_ops/onnx_ops_gpu_rocm_rdna3.json:
  a.) Earlier failing compilation, now passing inference:
    - onnx/node/generated/test_convtranspose_kernel_shape
    - onnx/node/generated/test_einsum_sum

b.) Earlier failing compilation, now passing compilation but failing
inference
    - onnx/node/generated/test_convtranspose_output_shape

  c.) Earlier passing but now failing inference
    - onnx/node/generated/test_scan_sum
    - onnx/node/generated/test_scan9_sum
    - onnx/node/generated/test_slice_default_axes

- For tests/external/iree-test-suites/onnx_ops/onnx_ops_gpu_vulkan.json:
  a.) Earlier failing compilation, now passing inference:
    - onnx/node/generated/test_convtranspose_kernel_shape

b.) Earlier failing compilation, now passing compilation but failing
inference
    - onnx/node/generated/test_convtranspose_output_shape
    - onnx/node/generated/test_einsum_sum

  c.) Earlier passing but now failing inference
    - onnx/node/generated/test_scan_sum
    - onnx/node/generated/test_scan9_sum

This commit
https://github.com/llvm/torch-mlir/commit/55ff110dc29cab7e2495ccdbec9a60512c29c665
is expected to fix the newly introduced failures which will be included
in IREE in the next Torch-MLIR bump.

---------

Signed-off-by: Vivek Khandelwal <vivekkhandelwal1424@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

ScottTodd committed:
Document new external ONNX model and linalg operator test suites. (#18819)

Progress on https://github.com/iree-org/iree-test-suites/issues/2 and
https://github.com/iree-org/iree-test-suites/issues/6 .

Signed-off-by: Elias Joseph <eljoseph@amd.com>

Commit:
disabled rebuild for testing purposes

Signed-off-by: Elias Joseph <eljoseph@amd.com>

nirvedhmeshram committed:
[ROCM][NFC] Add option to control SLP vectorization in llvm optimizations (#18865)

We keep SLP vectorization off because it can mask perf issues or create
regressions. However, on ROCM what we have noticed is that we are
hitting issues in several untested paths in the AMDGPU llvm backend
because we dont have SLP vectorization.
Here is an example of an issue that we wouldn't hit if SLP vectorization
was turned on
https://github.com/iree-org/iree/issues/18798

In this PR we are still keeping the exisiting behavior but provide a
flag to toggle it so that we can do the required benchmarking and
analysis.

Signed-off-by: Nirvedh <nirvedh@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

MaheshRavishankar committed:
Allow dynamic dimensions during folding of `tensor.expand_shape/collapse_shape` into `flow.dispatch.tensor.load/store`. (#18873)

This also cleans up the implementation of these patterns to avoid using
templated code that is hard to read/maintain.

---------

Signed-off-by: MaheshRavishankar <mahesh.ravishankar@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

stellaraccident committed:
Use FetchContent for both pybind11 and nanobind. (#18872)

This avoids various pinning problems and ensures that the version that
IREE specifies at the top of a build is used consistently throughout.
The nanobind incantation was taken from shortfin. The pybind11
incantation was adapted from [a
comment](https://github.com/pybind/pybind11/issues/2817#issuecomment-766392922)
and extended to use find_package integration.

The latter is sufficient for MLIR's pybind11 finding to delegate to the
one set at the top level. The MLIR code for finding pybind11 is ancient
and should be modernized to use FetchContent and find_package
integration with a pinned version. This would ensure consistent interop
with the rest of the ecosystem.

---------

Signed-off-by: Stella Laurenzo <stellaraccident@gmail.com>
Co-authored-by: Marius Brehler <marius.brehler@amd.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

Max191 committed:
[Codegen] Add pass to decompose pack unpack ops at dispatch boundaries (#18852)

This PR adds a wrapper pass around DecomposePackUnPackOps, which adds a
control function for decomposing only packs and unpacks whose reshapes
can be folded with dispatch tensor loads/stores.

This PR also removes the public pass constructor with a control
function, opting to use wrapper passes in place of constructing passes
with arbitrary control functions. This is better for creating simple bug
repros, since the control function is part of the pass.

---------

Signed-off-by: Max Dawkins <max.dawkins@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

benvanik committed:
Preserving `nosideeffects` on func.func -> util.func import. (#18882)

Signed-off-by: Elias Joseph <eljoseph@amd.com>

benvanik committed:
Emit an error when affinity analysis fails. (#18883)

Includes the flag I'd tell anyone to use if they filed a bug.

Fixes #18878.

Signed-off-by: Elias Joseph <eljoseph@amd.com>

Max191 committed:
[LLVMGPU] Use forall workgroup distribution in TileAndFuse pipeline (#18565)

This switches the TileAndFuse pipeline to use scf.forall distribution.
Using scf.forall distribution also requires some changes to the pass
ordering in the TileAndFuse pipeline, which is also handled by this PR:
1. The main difference is that PackToIntrinsics happens before workgroup
distribution. Otherwise, collapse_shape ops can end up at the end of the
workgroup forall, and an extra buffer is created.
2. Pack decomposition is now staged, with packs/unpacks at the function
boundaries being decomposed early before workgroup decomposition, and
the rest being decomposed after reduction tiling as before. This
prevents unpacks being fused into the workgroup forall and causing the
same problem as in (1).
3. `ConcretizeMmaShapes` now runs before workgroup tiling as well,
so the resulting collapse_shape on the multi_mma op result can be
propagated to the function boundary before any tiling. This is also to
avoid the same problem as in (1).

The lowering configs on the MMA path have also changed, since they now
need to account for inner tile sizes of packing.

depends on https://github.com/iree-org/iree/pull/18852

Signed-off-by: Max Dawkins <max.dawkins@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

IanWood1 committed:
Add conversions for 1x1 conv_2d to matmul (#18736)

Convert1X1FilterConv2DToMatmul: handle dynamic cases and conversion to `linalg.generic` representing a broadcasted batch matmul. This pass is kept because it is used in plugins.

GeneralizeLinalgNamedOps: generalize conv ops to `linalg.generic` ops when possible.

Converting more ops to linalg.generic ops allows for better reshape propagation and fusion opportunities. Also, removed Convert1X1FilterConv2DToMatmulPass from global optimization because generalize named ops would have already generalized any convolutions that were possible to convert.

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

bjacob committed:
Support 8-bit floats in the runtime (#18885)

There are 2 commits in this PR:
1. `math.h` conversions for 8-bit float types. This was previously sent
for review as https://github.com/iree-org/iree/pull/17989 but is now
folded into this PR.
2. General runtime/HAL support.

---------

Signed-off-by: Benoit Jacob <jacob.benoit.1@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

bjacob committed:
Support 8-bit floats in the compiler. (#18886)

This is a step in a series of PRs adding support for 8-bit flows. It's
sandwiched between https://github.com/iree-org/iree/pull/18885 and
subsequent PRs that will actually do something useful with this.

---------

Signed-off-by: Benoit Jacob <jacob.benoit.1@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

bjacob committed:
Support F8E5M2FNUZ MFMA on CDNA3 (#18887)

F8E4M3FNUZ was already there.

---------

Signed-off-by: Benoit Jacob <jacob.benoit.1@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

bjacob committed:
Add e2e tests for F8E5M2FNUZ and F8E4M3FNUZ data-tiled MFMA on CDNA3 (#18888)

Signed-off-by: Benoit Jacob <jacob.benoit.1@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

Commit:
added backup link to mnist test

Signed-off-by: Elias Joseph <eljoseph@amd.com>

bjacob committed:
Simplifications in e2e matmul tests (#18889)

Two commits:
1. Stop inferring `acc_type`. Require specifying it. Only a few tests
were relying on the inferrence.
2. Stop special-casing narrow float types (only using f32 as ABI type,
generating `arith.truncf` internally). This was only needed when these
narrow float types were not supported in the rest of IREE.

Signed-off-by: Benoit Jacob <jacob.benoit.1@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

nirvedhmeshram committed:
[GPU] Bail out in GPUReduceBankConflicts if we have collapse_shape user (#18863)

This is unsupported by upstream and can lead to a compiler error.
https://github.com/llvm/llvm-project/issues/112994

Progress towards: https://github.com/iree-org/iree/issues/18858

---------

Signed-off-by: Nirvedh <nirvedh@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

raikonenfnu committed:
[VectorLayout] Fix insertion of new constOp for non dominate issue. (#18894)

Main motivation of this patch is to resolve issue where we have the same
constOp being used by multiple operations.

But with a twist where first time the constOp needs a layout is on a op
that topologically comes after other ops that use constOp. This will
generate a copy of constOp in the location right before the latter op,
which is problematic because this constOp will be used by other ops
before it.

Previously for the test added we get this error:
```
within split at contraction_layout.mlir:1 offset :24:10: note: see current operation: %9 = "arith.addf"(%8, %6) <{fastmath = #arith.fastmath<none>}> : (vector<96x64xf16>, vector<96x64xf16>) -> vector<96x64xf16>
within split at contraction_layout.mlir:1 offset :22:19: error: operand #1 does not dominate this use
    %scaled_rhs = arith.mulf %read_1, %cst_1 : vector<96x64xf16>
```

While minor, this is also problematic because this error seem to stopped
layout analysis (but not fatally) S.T it fails to vector distribute in
some cases, making it hard to debug.

Signed-off-by: Stanley Winata <stanley.winata@amd.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

pashu123 committed:
[LLVMCPU] Enable tileDispatchUsingForall as default (#18777)

Signed-off-by: Elias Joseph <eljoseph@amd.com>

bjacob committed:
Add CDNA3 MFMA BF16 intrinsics. (#18892)

Signed-off-by: Benoit Jacob <jacob.benoit.1@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

Groverkss committed:
[LLVMGPU] Combine parallel and reduction padding in LLVMGPUPadAndVectorDistribute (#18771)

Since https://github.com/iree-org/iree/pull/18748 tensor.pad can be
fused in with tiling. This patch combines the parallel and reduction
padding passes into a single pass that pads at once, and the pads are
later fused during tiling.

Signed-off-by: Elias Joseph <eljoseph@amd.com>

nithinsubbiah committed:
[Dispatch Creation] Bubble up ExtractSliceOp with FillOp when the latter has multiple consumers (#18896)

Signed-off-by: nithinsubbiah <nithinsubbiah@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

Max191 committed:
[LLVMGPU][NFC] Create LLVMGPU pass for IGEMM (#18871)

This PR refactors the ConvolutionToIGEMM pass to a shared transform
function, and creates a new pass for LLVMGPU. This keeps the lowering
config details in LLVMGPU separate from the common pass, and removes the
need for passing a control function or config function in the pass
constructor. This is also a precursor to adding some more complex logic
in the control function for LLVMGPU, which will be added in a later PR.

---------

Signed-off-by: Max Dawkins <max.dawkins@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

IanWood1 committed:
[NFC] Update old naming from flow to dispatch creation (#18904)

Update naming from `Flow` -> `DispatchCreation` in Passes.cpp

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

Max191 committed:
[GPU] Support multiple contraction dims in MmaSchedules (#18720)

This adds support for multiple M, N, and K dims in problems when
deducing a GPUMMASchedule. The new heuristic is similar to the old one,
but works on pairs of M and N dims. For example:
```
tensor<M1xM0xK1xK0> * tensor<N1xN0xK1xK0> -> tensor<M1xN1xM0xN0>
```
This will try to distribute the seeded tile counts to `M0` and `N0`
(first attempting to distribute evenly, and then distributing to N
followed by N), and then distribute the residual counts to `M1` and
`N1`. The K tile counts will be partitioned to `K0` first, and then the
residual tile counts will be partitioned to `K1`.

This PR also updates the config selection logic for the TileAndFuse
pipeline to make use of the multiple contraction dimensions in mma
schedules.

---------

Signed-off-by: Max Dawkins <max.dawkins@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

MaheshRavishankar committed:
Pass to block dynamic dimensions of operands of `iree_linalg_ext.attention`. (#18874)

The use of `IntegerRangeAnalysis` and `IntegerDivisibilityAnalysis`
gives range and divisibility information for constants passed to the
dispatch. This can be used to infer the range and divisibility
information for all tensor values in the dispatch. This PR adds an
analysis to do this.

This analysis is then used to expand the dimensions of operands of the
attention operation that are dynamic, but are known to be divisible by
a compile-time static value. This gets the operations into a form that
can be compiled by the AMDGPU backend and target the mfma intrinsics.

Signed-off-by: MaheshRavishankar <mahesh.ravishankar@gmail.com>

---------

Signed-off-by: MaheshRavishankar <mravisha@amd.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

benvanik committed:
Adding build config for HSA runtime headers. (#18909)

Signed-off-by: Elias Joseph <eljoseph@amd.com>

benvanik committed:
Minor comment/style tweaks to the null HAL driver. (#18911)

Signed-off-by: Elias Joseph <eljoseph@amd.com>

benvanik committed:
Adding HAL semaphore support for statuses-as-failure-payloads. (#18912)

This allows an implementation to have a single atomic value for a
semaphore that encodes the user payload or an error payload that
optionally references an iree_status_t object. Implementations not using
the status feature can ignore it but must perform a
greater-than-or-equal check on `IREE_HAL_SEMAPHORE_FAILURE_VALUE`
instead of equality.

Signed-off-by: Elias Joseph <eljoseph@amd.com>

benvanik committed:
Modernizing iree_atomic_*. (#18910)

C11's _Generic lets us avoid the need for specifying the type in the
name and more closely match the C11 atomic syntax. This assumes that any
C compiler we have that goes down the disabled atomics path supports
_Generic (modern GCC, Clang, and MSVC all have for awhile).

This allows us to drop-in replace C11-style atomics (useful in the new
AMDGPU backend) and on MSVC will allow us to use their implementation
when it's ready (it's way better than the Interlocked solution we have
now).

Signed-off-by: Elias Joseph <eljoseph@amd.com>

IanWood1 committed:
Revert "[DispatchCreation] Extend multi-use producer fusion" (#18917)

The reverted commit does not handle when the "consumer" uses a value
defined above. See https://github.com/iree-org/iree/issues/18879 for the
original issue. This is causing issue with ~15 onnx models.

I have a PR (https://github.com/iree-org/iree/pull/18855) to fix this by
including values used in an ops region in the backwards slice, but It is
waiting on upstream changes to `getBackwardSlice`. Currently, the PR is
using a wrapper around `getBackwardSlice` to acheive the same effect,
but this will be updated once the upstream change lands
(https://github.com/llvm/llvm-project/pull/113478)

Reverts iree-org/iree#18551

---------

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

bangtianliu committed:
[VectorDistribution] Add vector distribution support multi-dim reduction with scalars (#18800)

Splitting https://github.com/iree-org/iree/pull/18519 into four patches.

Depends #18784

This is the second one, adding the corresponding layout analysis and
especially supporting the case where reduction is performed inside
scf.for operation.

Also, the relevant tests are added.

Since patch 2 includes changes from patch #18784, the necessary updates
from the first patch have also been included here.

---------

Signed-off-by: Bangtian Liu <liubangtian@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

Groverkss committed:
[LinalgExt] Generalize attribute setting for attention decomposition (#18780)

This PR teaches attention decomposition to set attributes for attention
matmuls by passing attribute dictionaries to
iree_linalg_ext.online_attention operation. This allows us to further
control codegen of matmuls (generally the root operations) after
decomposition (for example, setting lowering config on the decompose
matmuls).

Signed-off-by: Elias Joseph <eljoseph@amd.com>

Max191 committed:
[Codegen] Don't require full slice to decompose boundary pack and unpack ops (#18906)

This PR loosens the restrictions on decomposing boundary pack and unpack
ops. The current restriction is that the dispatch.tensor.load/store ops
are full slices, but this is not necessary for the current use case in
the TileAndFuse pipeline.

Instead, it is better for the time being to decompose non-padded
pack/unpack ops at function boundaries regardless of the
dispatch.tensor.load/store ops being full slices, because decomposing
such ops later on can cause issues with DPS. The DPS issues are tracked
in https://github.com/iree-org/iree/issues/18902, but we can loosen the
restrictions regardless, since it does not pose any issues to decompose
in such cases.

Signed-off-by: Max Dawkins <max.dawkins@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

benvanik committed:
Fixing missing parameters module fork_state vtable entry.

Signed-off-by: Elias Joseph <eljoseph@amd.com>

qedawkins committed:
Use workgroup_count_from_slice in Stream builtins (#18924)

`workgroup_count_from_dag_root` is planned to be replaced in the future
and is not supported by all codegen paths. Switch to
`workgroup_count_from_slice`.

Signed-off-by: Elias Joseph <eljoseph@amd.com>

stellaraccident committed:
Enable the MLIR debug actions CL options in the compiler driver. (#18928)

Signed-off-by: Stella Laurenzo <stellaraccident@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

IanWood1 committed:
[DispatchCreation] Run preprocessing before elementwise fusion (#18920)

I think it makes sense to run `FusionPreprocessingPass` before
`ElementwiseOpFusionPass` because it helps put the IR in a better state
for fusion (e.g. interchanging `linalg.generic` indexing maps). But
also, reshapes have been propagated to the edges of the program, which
allows the `GatherFusionPattern` to be more effective.

Fixes compilation error from
https://github.com/iree-org/iree/issues/17226#issuecomment-2441200369.

---------

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

qedawkins committed:
[ROCM] Add flag to enable GlobalISel (#18922)

This exposes a flag to control the use of GlobalISel in the ROCm LLVM backend. There are observed correctness issues with SelectionDAG and the plan is to turn this on by default once all issues have been triaged.

Signed-off-by: Elias Joseph <eljoseph@amd.com>

marbre committed:
Work around missing pybind in BYO LLVM build (#18916)

Installs pybind via pip to work around #18884.

Closes #18884.

Signed-off-by: Elias Joseph <eljoseph@amd.com>

IanWood1 committed:
Revert "[DispatchCreation] Run preprocessing before..." (#18934)

This PR got merged before I was able to resolve the perf regressions in
VAE decode on MI250. See @ScottTodd's comment on the original PR.
I need time to resolve the regressions but this can be relanded once
resolved

Reverts iree-org/iree#18920

Signed-off-by: Elias Joseph <eljoseph@amd.com>

Groverkss committed:
[LinalgExt] Implement AggregateOpInterface for AttentionOp (#18890)

- Adds AggregateOpInterface for AttentionOp
- Move all aggregate interface tests to IR/test/decompose_aggregate_op

Signed-off-by: Elias Joseph <eljoseph@amd.com>

Groverkss committed:
[GPU] Do not treat pad as a tilable producer for operand promotion (#18918)

PadOp doesn't have an implementation for deriving thread configuration
from derived_thread_config, so ignore promoting it until an
implementation is added.

Signed-off-by: Elias Joseph <eljoseph@amd.com>

benvanik committed:
Adding 'amdgpu' target device and flatbuffer for HAL executables. (#18933)

The schema may change as the branch gets closer to merging but the
refactoring in the compiler for serializing multiple ABIs will remain
the same.

Signed-off-by: Elias Joseph <eljoseph@amd.com>

Commit:
attempting removal of msvc install based on new dockerfile

Signed-off-by: Elias Joseph <eljoseph@amd.com>

benvanik committed:
Enabling linking in the ROCM/CUDA compiler targets. (#18936)

This does exactly what the LLVMCPU side does - which is bad for compile
time (serializes LLVM codegen) but much better for runtime. Future
improvements should move LLVM codegen to the linking phase so it can
happen in parallel and then perform the linking using LLVM's linker
(each executable turned into a .o and then combined into a .so, or
last-level bitcode if then we just want serialization to be bitcode to
machine code). This is definitely a compile-time regression but we can't
keep pessimizing runtime.

Signed-off-by: Elias Joseph <eljoseph@amd.com>

stellaraccident committed:
Add integer range inference to hal.buffer_view.dim and rank ops. (#18943)

This matches that default range behavior of runtime dimensions we get
from frontends.

---------

Signed-off-by: Stella Laurenzo <stellaraccident@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

benvanik committed:
Registering the ROCDL dialect in init_mlir_dialects. (#18944)

We're emitting those ops now (directly or indirectly) and in order to
parse IR that contains them (from e.g. resuming
--compile-to=executable-targets) we must have all dialects registered.

Signed-off-by: Elias Joseph <eljoseph@amd.com>

c-rhodes committed:
[GlobalOpt] Prevent fusing transposed extend in RaiseSpecialOps (#18901)

NamedImplicitCastOpConversion pattern is incorrectly fusing transposed
element-wise extend into Linalg op.

---------

Signed-off-by: Cullen Rhodes <cullen.rhodes@arm.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

benvanik committed:
Switching VM's EraseUnusedCallOp pattern to a pass. (#18950)

This lets it use a symbol table to speed things up _a lot_.

Includes various fixes found during testing.

Signed-off-by: Elias Joseph <eljoseph@amd.com>

MaheshRavishankar committed:
Add `ReifyRankedShapedTypeOpInterface` to `hal.interface.binding.subspan` (#18946)

Fixes #18942

Signed-off-by: MaheshRavishankar <mahesh.ravishankar@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

bjacob committed:
GPU data tiling: reimplement getConcreteMFMALayout (#18953)

This `getConcreteMFMALayout` function had its own independent database
of layout details for MMA intrinsics. Instead, reimplement it using the
tile swizzle infra (which gets its information about layout of MMA
intrinsics from `getSingleSubgroupLayout`).

Since this function is exercised by lit tests covering RDNA3 WMMA
intrinsics with zero strides, this forced solving the issue about that
in GPUTileSwizzleUtils.cpp, a blocker for GPU data tiling on RDNA3.

Signed-off-by: Benoit Jacob <jacob.benoit.1@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

benvanik committed:
Adding a flag to force indirect command buffers on in non-reusable cases. (#18945)

Includes various fixes found during testing.

Signed-off-by: Elias Joseph <eljoseph@amd.com>

nirvedhmeshram committed:
[ROCM] Turn on SLP vectorization (#18949)

After doing review on the benchmarks we see that this is not causing any
model regression and we also did a benchmarking suite of attention
kernels for which in the past we thought this could cause a regression.
However, we didnt find any significant perf change. See
[here](https://docs.google.com/spreadsheets/d/102hYwdOGehmi_HhnLxHevwAVzMraKgKEjdTdUlJ4_TU/edit?usp=sharing)
As this is needed for correctness issue found in
https://github.com/iree-org/iree/issues/18798 we are turning on the SLP
vectorizer.

Signed-off-by: Nirvedh <nirvedh@gmail.com>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

IanWood1 committed:
Propagate reshapes through generics with reduction iterators (#18857)

Removes the constraint in `BubbleUpExpandShapes` that prevents moving
tensor reshape ops through reduction `linalg.generic` ops. This has the
benefit of increasing the dimensionality of reduction ops (more fusion
opportunities) as well as increasing the chance these ops will be moved
to the edge of the program.

Closes https://github.com/iree-org/iree/issues/18854

---------

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>
Signed-off-by: Elias Joseph <eljoseph@amd.com>

Groverkss committed:
[LLVMGPU] Use flat workgroup sizes in vector distribution (#18947)

There is no reason for vector distribution to use this weird workgroup
split. Use flat workgroup sizes in KernelConfig instead.

Signed-off-by: Elias Joseph <eljoseph@amd.com>

qedawkins committed:
[Flow] Add pattern to canonicalize away full tensor.insert_slice ops (#18941)

Additionally drops the pad combining pattern from this pass because it
was upstreamed.

Signed-off-by: Elias Joseph <eljoseph@amd.com>

Groverkss committed:
[Codegen][LLVMGPU] Set global read layouts at linalg level (#18860)

- Operand promotion is now done the same way as TileAndFuse pipeline, by
reading promote_operands config from lowering_config.
- Moves global read layout setting to LLVMGPUConfigureTensorLayouts,
from LLVMGPUConfigureVectorLayouts pass, anchoring based on lowering
config.

These changes by side effect allow setting layouts on gathers in
VectorDistribute pipeline.

Signed-off-by: Elias Joseph <eljoseph@amd.com>

benvanik committed:
Improving VM conversion performance. (#18957)

The major change here is using a precomputed import table in VM
conversion patterns. This removes the symbol lookup that was happening
on each call. In models with 100k calls to imports this speeds things up
a lot.

Also squashed a few more perf issues involving symbol lookups while
profiling and made some passes that could nest on function-like ops do
so.

These changes drop VM translation of the 405b model from 3.5mins to
~1.5min. Disabling verification (`-verify-each=0` to iree-opt or
`-verify=false` to iree-compile) takes it to 1min.

Remaining work is mostly around parallelizing some passes that are not
trivially parallelizable (FoldGlobals, DropUnusedCalls, etc) and
parallelizing some analysis (Explorer global init, call graph walking)
that tends to get real expensive when there are 250k calls and 500k ops.
Any place that does a symbol use walk is going to suffer. Many of these
fixes are in our code but there's several upstream components that fall
over with this amount of IR (CallGraph, DataFlowSolver, the verifier,
etc).

Signed-off-by: Elias Joseph <eljoseph@amd.com>
